### PR TITLE
Nrf9160 soc

### PR DESCRIPTION
**Pull request by ioannisg**
This PR contributes the nRF9160 SOC definition and the basic init configuration

### CODE REVIEW

File: soc/arm/Kconfig [hunk 1]
```diff
@@ -20,3 +20,12 @@ config CPU_HAS_ARM_SAU
 	select CPU_HAS_TEE
 	help
 	  MCU implements the ARM Security Attribution Unit (SAU).
+
+config CPU_HAS_NRF_IDAU
```
Comment (ioannisg) [inline soc/arm/Kconfig]: FYI @agross-linaro : this option represents the Nordic IDAU.

File: soc/arm/Kconfig [hunk 2]
```diff
@@ -20,3 +20,12 @@ config CPU_HAS_ARM_SAU
...
+	bool
+	# Omit prompt to signify "hidden" option
+	depends on SOC_SERIES_NRF91X
+	select CPU_HAS_TEE
+	help
+	  MCU implements the nRF (Implementation-defined)
```
Comment (pizi-nordic) [inline soc/arm/Kconfig]: `(Implementation defined)` after nRF does not look well.
What about "MCU implements Nordic flavour of Security... (IDAU)."

File: soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.nrf9160_SICA [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+
+if SOC_NRF9160_SICA
+
+config SOC
+	string
+	default "nRF9160_SICA"
```
Comment (pizi-nordic) [inline soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.nrf9160_SICA]: Why "SICA"?

File: soc/arm/nordic_nrf/nrf91/Kconfig.series [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	bool "Nordic Semiconductor nRF91 series MCU"
+	select CPU_CORTEX_M33
+	select CPU_HAS_ARM_MPU
+	select CPU_HAS_NRF_IDAU
+	select CPU_HAS_FPU
+	select ARMV8_M_DSP
+	select SOC_FAMILY_NRF
+	select NRF_RTC_TIMER
+	select CLOCK_CONTROL
+	select CLOCK_CONTROL_NRF5
```
Comment (pizi-nordic) [inline soc/arm/nordic_nrf/nrf91/Kconfig.series]: Why you need CLOCK_CONTROL when you are requesting CLOCK_CONTROL_NRF5?
Is there a missing requirement in CLOCK_CONTROL_NRF5?

File: soc/arm/nordic_nrf/nrf91/dts_fixup.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* SoC level DTS fixup file */
+
+#define DT_NUM_IRQ_PRIO_BITS        DT_ARM_V8M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
```
Comment (pizi-nordic) [inline soc/arm/nordic_nrf/nrf91/dts_fixup.h]: It looks that there is some alignment problem on this line.

File: soc/arm/nordic_nrf/nrf91/mpu_regions.c [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/slist.h>
+#include <arch/arm/cortex_m/mpu/arm_mpu.h>
+
+
+#define XICR_BASE	0x10000000
```
Comment (pizi-nordic) [inline soc/arm/nordic_nrf/nrf91/mpu_regions.c]: Do we have any header with these value defined?
Why we are not fetching these addresses from dts?
Why we have defines which are not used?

File: soc/arm/nordic_nrf/nrf91/soc.c [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+#include <logging/log.h>
+LOG_MODULE_REGISTER(soc);
+
+static int nordicsemi_nrf91_init(struct device *arg)
+{
+	u32_t key;
+
+	ARG_UNUSED(arg);
+
+	key = irq_lock();
```
Comment (pizi-nordic) [inline soc/arm/nordic_nrf/nrf91/soc.c]: Are we really expect and an interrupt on SoC init level?

**Review by agross-oss — COMMENTED** [review]
as a whole it looks ok.  I'd like to see the CLOCK_CONTROL fixed.  Aside from that, the rest of the comments are nits.

File: soc/arm/nordic_nrf/nrf91/soc.c [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+#endif
+
+#if defined(CONFIG_SOC_NRF9160)
+#include <system_nrf9160.h>
+#else
+#error "Unknown SoC."
+#endif
+
+#include <nrf.h>
+#include <hal/nrf_power.h>
```
Comment (anangl) [inline soc/arm/nordic_nrf/nrf91/soc.c]: These two inclusions do not seem to be needed (`<nrf.h>` is included by `<nrfx.h>`, and POWER HAL is not used here).

File: soc/arm/nordic_nrf/nrf91/soc.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * @file SoC configuration macros for the Nordic Semiconductor nRF91 family processors.
+ */
+
+#ifndef _NORDICSEMI_NRF91_SOC_H_
+#define _NORDICSEMI_NRF91_SOC_H_
+
+#ifndef _ASMLANGUAGE
+
+#include <nrf_common.h>
```
Comment (anangl) [inline soc/arm/nordic_nrf/nrf91/soc.h]: This is needless.

File: soc/arm/nordic_nrf/nrf91/dts_fixup.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+#define DT_ADC_0_NAME                   DT_NORDIC_NRF_SAADC_ADC_0_LABE
```
Comment (anangl) [inline soc/arm/nordic_nrf/nrf91/dts_fixup.h]: One character (`L`) at the end of the line got eaten somehow.

File: soc/arm/nordic_nrf/nrf91/soc.c [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+#include <soc/nrfx_coredep.h>
+
+#ifdef CONFIG_RUNTIME_NMI
+extern void _NmiInit(void);
+#define NMI_INIT() _NmiInit()
+#else
+#define NMI_INIT()
+#endif
+
+#if defined(CONFIG_SOC_NRF9160)
```
Comment (lemrey) [inline soc/arm/nordic_nrf/nrf91/soc.c]: Why is this necessary?

File: soc/arm/nordic_nrf/nrf91/soc.c [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+#include <system_nrf9160.h>
+#else
+#error "Unknown SoC."
+#endif
+
+#include <nrf.h>
+#include <hal/nrf_power.h>
+
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
```
Comment (lemrey) [inline soc/arm/nordic_nrf/nrf91/soc.c]: I'd suggest to pass this to the `LOG_MODULE_REGISTER` macro directly.

File: soc/arm/nordic_nrf/nrf91/soc.c [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+#include <logging/log.h>
```
Comment (lemrey) [inline soc/arm/nordic_nrf/nrf91/soc.c]: I think this can be moved next to the other inlcudes.

File: soc/arm/nordic_nrf/nrf91/dts_fixup.h [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+#define DT_UART_0_NAME			DT_NORDIC_NRF_UARTE_UART_0_LABEL
+#define DT_UART_1_NAME			DT_NORDIC_NRF_UARTE_UART_1_LABEL
+#define DT_UART_2_NAME			DT_NORDIC_NRF_UARTE_UART_2_LABEL
+#define DT_UART_3_NAME			DT_NORDIC_NRF_UARTE_UART_3_LABEL
+
+#define DT_FLASH_DEV_NAME \
+	DT_NRF_NRF91_FLASH_CONTROLLER_FLASH_CONTROLLER_LABEL
```
Comment (lemrey) [inline soc/arm/nordic_nrf/nrf91/dts_fixup.h]: I am unable to find `DT_NRF_NRF91_FLASH_CONTROLLER_FLASH_CONTROLLER_LABEL` in a `generated_dts_board.h` file for a project built for `nRF9160_pca10090`. Could you double check that this is correct?

File: soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.nrf9160_SICA (hunk 3 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.nrf9160_SICA]: @pizi-nordic  This is the code-name for current flavor of nrf9160

Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.nrf9160_SICA]: See the product options of nRF9160: https://www.nordicsemi.com/DocLib/Content/Product_Spec/nRF9160/latest/nRF9160_html5_keyfeatures

File: soc/arm/Kconfig (hunk 2 above)
Comment (ioannisg) [reply soc/arm/Kconfig]: new text:

> MCU implements the nRF (vendor-specific) Security Attribution Unit.
 (IDAU: "Implementation-Defined Attribution Unit", in accordance with
  ARM terminology).

Hope this clarifies it better, @pizi-nordic

File: soc/arm/nordic_nrf/nrf91/Kconfig.series (hunk 4 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/Kconfig.series]: nRF51 and nRF52 SOCs (and simulated nRF boards) select both Kconfig options - this is perhaps a bit problematic, but cannot be solved under this PR. We should need a separate pull request for that. Would you be OK with this? I can track it under #12199

File: soc/arm/nordic_nrf/nrf91/Kconfig.soc [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+	select HAS_HW_NRF_TWIM3
+	select HAS_HW_NRF_TWIS0
+	select HAS_HW_NRF_TWIS1
+	select HAS_HW_NRF_TWIS2
+	select HAS_HW_NRF_TWIS3
+	select HAS_HW_NRF_UARTE0
+	select HAS_HW_NRF_UARTE1
+	select HAS_HW_NRF_UARTE2
+	select HAS_HW_NRF_UARTE3
+	select HAS_HW_NRF_WDT
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/Kconfig.soc]: I think so. For now, we keep compliance with nRF52 and 51

File: soc/arm/nordic_nrf/nrf91/dts_fixup.h (hunk 10 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/dts_fixup.h]: Fixing it, thanks.

File: soc/arm/nordic_nrf/nrf91/dts_fixup.h (hunk 5 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/dts_fixup.h]: Thanks, fixing it.

File: soc/arm/nordic_nrf/nrf91/Kconfig.series (hunk 4 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/Kconfig.series]: @pizi-nordic @agross-linaro I did some investigation here:
- In some SOC/SOC_SERIES (e.g. NXP),the vendor-specific option is defined in the "defconfig" file, and set to "default=y" if CLOCK_CONTROL is selected already (at SOC level)
- In several SOC/SOC_SERIES (e.g. STM32) CLOCK_CONTROL_<vendor-specific> is _selected_ conditionally,if CLOCK_CONTROL is selected (normally) at the BOARD level.
For example `select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL`
- For Nordic nRFx, CLOCK_CONTROL is directly selected at the SOC_SERIES_NRFxx. That is why we only need:  `select CLOCK_CONTROL_NRF5`, i.e. we can skip the conditional (if CLOCK_CONTROL)

Conclusions:
1. It is the other way @agross-linaro , i.e. CLOCK_CONTROL selects the vendor-specific option
2. I can make this look like the STM32 case, i.e. `select CLOCK_CONTROL_NRF5 if CLOCK_CONTROL` but it is not really needed since CLOCK_CONTROL is selected under the same symbol. So I would rather leave it as is.


---

We _do_ have an issue with naming, here, since CLOCK_CONTROL_NRF5 is "used" in nRF91 series. I think we can  simply rename CLOCK_CONTRO_NRF5 --> CLOCK_CONTROL_NRF. I think I will open a separate PR for this.

File: soc/arm/nordic_nrf/nrf91/soc.c (hunk 7 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/soc.c]: I agree that it can happen.

File: soc/arm/nordic_nrf/nrf91/mpu_regions.c (hunk 6 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/mpu_regions.c]: I think this can be cleaned-up. These defines are in the .c file, thus, cannot be used outside this source file. I added them to be consistent with nRF52 but I see that they are neither used in nRF52. Similar defines are added in some NXP and ST SOCs. I suggest open a PR to clean them up, since they are not used.

Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/mpu_regions.c]: Removed, already, from this file.

File: soc/arm/nordic_nrf/nrf91/soc.c (hunk 8 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/soc.c]: OK, removed.
Note that the inclusion of nrf.h is present in the nrf51,51 soc.c files, too, so we need to clean this up from those, too.

File: soc/arm/nordic_nrf/nrf91/soc.h (hunk 9 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/soc.h]: Removed.
Again, note that this inclusion is present in nrf51, and 52 soc.h headers.

File: soc/arm/nordic_nrf/nrf91/soc.c (hunk 11 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/soc.c]: This keeps consistency with nRF52 soc.c, where there are multiple SOC variants. In the future we might have multiple nRF91 SOC variants, as well.

File: soc/arm/nordic_nrf/nrf91/soc.c (hunk 13 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/soc.c]: Will do. Note that the same needs to be done in nrf51 and 52 soc. files.

File: soc/arm/nordic_nrf/nrf91/soc.c (hunk 12 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/soc.c]: Is there any reason to do this differently from what we do in nRF51 and nRF52 soc.c?

File: soc/arm/nordic_nrf/nrf91/dts_fixup.h (hunk 14 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/dts_fixup.h]: This is a nice comment, thanks. So, this has been incorrect in downstream. where the fixup was defining `DT_FLASH_DEV_NAME `as `DT_NRF_NRF52_FLASH_CONTROLLER_4001E000_LABEL`. This is incorrect, not only for the "52" in-fix in the name, but also because of wrong address label - it may be 4001E000 or 5001E000, depending on security domain.

This has been already fixed in upstream, after merging the nrf91 DTS support. Therefore, we will now be able to generate the correct define:  	`DT_NRF_NRF91_FLASH_CONTROLLER_FLASH_CONTROLLER_LABEL `

**Comment by ioannisg** [comment]
@lemrey @pizi-nordic @anangl , I have addressed your comments, please, re-visit this review. 

[only issue remaining is the Ksymbols for CLOCK_CONTROL]

File: soc/arm/nordic_nrf/nrf91/soc.c (hunk 12 above)
Comment (pizi-nordic) [reply soc/arm/nordic_nrf/nrf91/soc.c]: The logger was (not so) recently updated to support simpler registration scheme. We should promote it :)

**Review by pizi-nordic — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Review by lemrey — APPROVED** [review]
LGTM

**Review by carlescufi — APPROVED** [review]
LGTM, thanks!

File: soc/arm/nordic_nrf/nrf91/Kconfig.series (hunk 4 above)
Comment (carlescufi) [reply soc/arm/nordic_nrf/nrf91/Kconfig.series]: >We do have an issue with naming, here, since CLOCK_CONTROL_NRF5 is "used" in nRF91 series. I think we can simply rename CLOCK_CONTRO_NRF5 --> CLOCK_CONTROL_NRF. I think I will open a separate PR for this.

Yes, this is the correct approach in my opinion. Let's go for this.

**Comment by carlescufi** [comment]
@agross-linaro @galak I think we can merge this one and rework the clock control issue in a follow-up PR that @ioannisg is working on.

**Comment by ioannisg** [comment]
> 
> 
> @agross-linaro @galak I think we can merge this one and rework the clock control issue in a follow-up PR that @ioannisg is working on.

@carlescufi @agross-linaro @galak, the refactoring of CLOCK_CONTROL_NRF symbol is already tracked as sub-task in #12199

**Comment by nashif** [comment]
What boards use this SoC? Can we submit a user in the same PR?

File: soc/arm/nordic_nrf/nrf91/Kconfig.series (hunk 4 above)
Comment (ioannisg) [reply soc/arm/nordic_nrf/nrf91/Kconfig.series]: @agross-linaro the CLOCK_CONTROL_NRF5 has been renamed to CLOCK_CONTROL_NRF, I suppose this issue could be resolved?

**Comment by ioannisg** [comment]
@galak this is now rebased over the latest changes in nordic dts
